deploy/stats.go

```diff
@@ -25,31 +132,157 @@ type StatsEntry struct {
 	UserTime float64 `json:"usertime"`
 }
 
-type ClientMsgStats struct {
-	File        string    `json:"file"`
-	Type        string    `json:"type"`
-	Buckets     []float64 `json:"buck,omitempty"`
-	RoundsAfter []float64 `json:"roundsAfter,omitempty"`
-	Times       []float64 `json:"times,omitempty"`
+// General interface to have each app have its own statistics displayed
+type Stats interface {
+	// Tell the stats to write on this specific writer
+	// user responsability to close it after if needed
+	// but it could be a net.Conn or whatever !
+	// Call this before ServerCSV* or ClientCSV*
+	WriteTo(w io.Writer)
+	ServerCSVHeader() error
+	ServerCSV() error
+	// had to keep client stats tight together because
+	// of the coll_stamp server's stats that needs the rate from the client
+	ClientCSV() error
+	ClientCSVHeader() error
+
+	// incoporate the Entry into theses stats
+	// i is the "Index" of this entry (i.e. # times we have added entries)
+	AddEntry(e Entry) error
+	// Valid tells is the stats received some real data and is not
+	// empty or full of garbage
+	Valid() bool
+}
+
+// statistics about the shamir_sign app
+type ShamirStats struct {
+	// the writer to write the stats
+	Writer io.Writer
+	// number of hosts
+	NHosts int
+
+	// times for the rounds
+	round StreamStats
+	// times for the setup
+	setup StreamStats
+
+	SysTime  float64
+	UserTime float64
+}
+
+func (s *ShamirStats) WriteTo(w io.Writer) {
+	s.Writer = w
+}
+
+// Return the CSV header of theses stats.
+// Could be implemented using reflection for automatic detection later .. ?
+func (s *ShamirStats) ServerCSVHeader() error {
+	_, err := fmt.Fprintf(s.Writer, "Hosts, %s, %s, user, system\n", s.round.Header("round_"), s.setup.Header("setup_"))
+	return err
+}
+
+func (s *ShamirStats) ServerCSV() error {
+	_, err := fmt.Fprintf(s.Writer, "%d, %s, %s, %f, %f\n",
+		s.NHosts,
+		s.round.String(),
+		s.setup.String(),
+		s.UserTime/1e9,
+		s.SysTime/1e9)
+	return err
+}
+
+func (s *ShamirStats) ClientCSVHeader() error {
+	return nil
+}
+func (s *ShamirStats) ClientCSV() error {
+	return nil
+}
+
+// Add an entry to the global stats
+func (s *ShamirStats) AddEntry(e Entry) error {
+	switch t := e.(type) {
+	// the entry is a ShamirEntry !
+	case ShamirEntry:
+		st := e.(ShamirEntry)
+		// is it about the Round , or the setup
+		if st.Type == ShamirRoundType {
+			s.round.Update(st.Time)
+		} else if st.Type == ShamirSetupType {
+			s.setup.Update(st.Time)
+		} else {
+			dbg.Fatal("Received unknown shamir entry : ", st.Type)
+		}
+	case SysEntry:
+		st := e.(SysEntry)
+		s.SysTime = st.SysTime
+		s.UserTime = st.UserTime
+	default:
+		dbg.Fatal("Received unknown entry type : ", t)
+	}
+	return nil
+}
+
+// basic check to see if we got somme real data
+func (s *ShamirStats) Valid() bool {
+	return s.round.Avg() > 0.0 && s.setup.Avg() > 0.0
 }
 
-type RunStats struct {
+// Average all these stats
+func shamirStatsAverage(stats ...Stats) (Stats, error) {
```

Comment:
I bet it wouldn't be hard to create a completely generic version of this function, which takes a varargs-list of interface{}, which it expects to be a slice of structs of float values, and creates/returns a tuple of structs of the same type containing the averages, mins, maxes, std devs, etc, of each field across all the input structs.  Then that could be reused for whatever application- or context-specific stats we want to deal with in any situation.  For example, the signature might be:

func SummarizeStats(stats ...interface{}) (min, max, avg, stddev interface{}) { ...}

Not sure it's necessarily worth the trouble, but something to consider anyway. :)